### PR TITLE
feat(mpc): add mpc transcoding template and tempate group resource

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

add mpc transcoding template and tempate group resource

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes NONE

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new resource huaweicloud_mpc_transcoding_template
2. add new resource huaweicloud_mpc_transcoding_template_group
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/mpc' TESTARGS='-run TestAccMPCTranscodingTemplate'==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mpc -v -run TestAccMPCTranscodingTemplate -timeout 360m -parallel 4
=== RUN   TestAccMPCTranscodingTemplateGroup_basic
=== PAUSE TestAccMPCTranscodingTemplateGroup_basic
=== RUN   TestAccMPCTranscodingTemplate_basic
=== PAUSE TestAccMPCTranscodingTemplate_basic
=== CONT  TestAccMPCTranscodingTemplateGroup_basic
=== CONT  TestAccMPCTranscodingTemplate_basic
--- PASS: TestAccMPCTranscodingTemplate_basic (24.59s)
--- PASS: TestAccMPCTranscodingTemplateGroup_basic (26.69s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mpc       26.783s
```

### CODE REVIEW

File: docs/resources/mpc_transcoding_template.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+* `ref_frames_count` - (Optional, Int) Specifies the maximum reference frames. When the `codec` is H.264,
+  the vaule rang is 1 to 8. When the `codec` is H.265, the value is fixed at 4. The default value is 4. The unit is frame.
+
+* `max_iframes_interval` - (Optional, Int) Specifies the mximum I-frame interval. The value ranges from 2 to 10.
+  The default value is 5. The unit is second.
+
+* `bframes_count` - (Optional, Int) Specifies the maximum number of B-frames.
+  The vaule range is  0 to 7, and the default value is 4. The unit is frame.
+
+* `frame_rate` - (Optional, Int) Specifies the frame rate. Its value is 0 or an integer ranging from 5 to 30.
```
Comment (Lance52259) [inline docs/resources/mpc_transcoding_template.md]: s/5 to 30/5 to 60

File: docs/resources/mpc_transcoding_template.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+* `output_policy` - (Optional, String) Specifies the output policy. Possible values are **discard** and **transcode**.
+  The default value is transcode.
+
+<a name="video_object"></a>
+The `video` block supports:
+
+* `output_policy` - (Optional, String) Specifies the output policy. Possible values are **discard** and **transcode**.
+  The default value is transcode.
+
```
Comment (Lance52259) [inline docs/resources/mpc_transcoding_template.md:117]: pls add a note to show more details:
+ If output_policy is set to discard in video parameters and transcode in audio parameters, only audio is output.
+ If output_policy is set to transcode in video parameters and discard in audio parameters, only video is output.
+ output_policy cannot be set to discard in video and audio parameters at the same time.
+ If output_policy is set to transcode in video and video parameters at the same time, video and video are output.

File: docs/resources/mpc_transcoding_template_group.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+* `ref_frames_count` - (Optional, Int) Specifies the maximum reference frames. When the `codec` is H.264,
+  the vaule rang is 1 to 8. When the `codec` is H.265, the value is fixed at 4. The default value is 4. The unit is frame.
+
+* `max_iframes_interval` - (Optional, Int) Specifies the mximum I-frame interval. The value ranges from 2 to 10.
+  The default value is 5. The unit is second.
+
+* `bframes_count` - (Optional, Int) Specifies the maximum number of B-frames.
+  The vaule range is  0 to 7, and the default value is 4. The unit is frame.
+
+* `frame_rate` - (Optional, Int) Specifies the frame rate. Its value is 0 or an integer ranging from 5 to 30.
```
Comment (Lance52259) [inline docs/resources/mpc_transcoding_template_group.md]: ditto.

File: huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	templateGroupList := *resp.TemplateGroupList
+
+	if len(templateGroupList) == 0 {
+		return nil, fmt.Errorf("unable to retrieve MPC transcoding template group: %s", state.Primary.ID)
+	}
+	templateGroup := templateGroupList[0]
+
+	return templateGroup, nil
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go]: `return templateGroupList[0], nil` just ok.

File: huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    output_policy        = "transcode"
+    preset               = 1
+    profile              = 3
+    ref_frames_count     = 4
+  }
+
+  videos {
+    width   = 3840
+    height  = 2160
+    bitrate = 0
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go:167]: The default value also suggests checking, which can be checked by not writing it in the script and using `resource.TestCheckResourceAttr(resourceName, "videos.0.bitrate", "0"),`.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	mpc "github.com/huaweicloud/huaweicloud-sdk-go-v3/services/mpc/v1/model"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/fmtp"
+)
+
+func ResourceMPCTranscodingTemplate() *schema.Resource {
```
Comment (Lance52259) [inline huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: `MPC` is redundant.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+	}
+
+	templateList := *resp.TemplateArray
+	template := templateList[0].Template
+	if template == nil {
+		log.Printf("unable to retrieve MPC transcoding template: %s", d.Id())
+		d.SetId("")
+		return nil
+	}
+
```
Comment (Lance52259) [inline huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: If the length of the returned result is 0, panic will be triggered: `invalid array index 0`.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+	return &outputPolicyOpts
+}
+
+func resourceMPCTranscodingTemplateCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.HcMpcV1Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating MPC client : %s", err)
```
Comment (Lance52259) [inline huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go:310]: suggest to add version information to facilitate quick positioning of API problems when errors occur.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+
+func flattenAudio(audio *mpc.Audio) []map[string]interface{} {
+	if audio == nil {
+		return nil
+	}
+	audioResult := make(map[string]interface{})
+	audioResult["codec"] = audio.Codec
+	audioResult["sample_rate"] = audio.SampleRate
+	audioResult["channels"] = audio.Channels
+	audioResult["bitrate"] = audio.Bitrate
```
Comment (Lance52259) [inline huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]:
```
audioResult := map[string]interface{}{
	"codec":       audio.Codec,
	"sample_rate": audio.SampleRate,
	"channels":    audio.Channels,
	"bitrate":     audio.Bitrate,
}
```
It's easier to read.
Code lines are clearly identified even when panic is triggered.

File: docs/resources/mpc_transcoding_template.md (hunk 1 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template.md]: the doc is right, validatefunc not correct, fixed

File: docs/resources/mpc_transcoding_template_group.md (hunk 3 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template_group.md]: fixed

File: docs/resources/mpc_transcoding_template.md (hunk 2 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template.md:117]: fixed

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: There is service name in most of our resource functions. I don't think this need to be change.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: I have tested it, the array is never empty.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 8 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go:310]: There is only one version of MPC api for now, so I don't think the version information is needed

File: huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go (hunk 4 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go]: sure

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 9 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: fixed

File: huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go (hunk 5 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/mpc/resource_huaweicloud_mpc_transcoding_template_group_test.go:167]: no need to change

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: I disagree, you will specify `mpc` twice if you refernece in the other packages: `mpc.ResourceMPCTranscodingTemplate()`.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go:366]: It's a risk, we cannot guarantee that future modifications on the service side will cause this panic to occur.
suggest to add a pre-check.

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: fair enough

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/mpc' TESTARGS='-run TestAcc'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mpc -v -run TestAcc -timeout 360m -parallel 4
=== RUN   TestAccTranscodingTemplateGroup_basic
=== PAUSE TestAccTranscodingTemplateGroup_basic
=== RUN   TestAccTranscodingTemplate_basic
=== PAUSE TestAccTranscodingTemplate_basic
=== CONT  TestAccTranscodingTemplateGroup_basic
=== CONT  TestAccTranscodingTemplate_basic
--- PASS: TestAccTranscodingTemplate_basic (23.44s)
--- PASS: TestAccTranscodingTemplateGroup_basic (25.74s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mpc       25.842s
```

File: docs/resources/mpc_transcoding_template.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+  + **1**: HLS
+  + **2**: DASH
+  + **3**: HLS+DASH
+  + **4**: MP4
+  + **5**: MP3
+  + **6**: ADTS
+
+-> If pack_type is set to 5 or 6, do not set video parameters.
+
+* `pvc` - (Optional, Bool) Specifies Whether to enable low bitrate HD. The default value is false.
```
Comment (niuzhenguo) [inline docs/resources/mpc_transcoding_template.md]: how about naming this to low_bitrate_hd as `Live` to keep consistent.

File: docs/resources/mpc_transcoding_template_group.md [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String, ForceNew) The region in which to create the transcoding template group resource. If omitted,
+  the provider-level region will be used. Changing this creates a new resource.
+
+* `name` - (Required, String) Specifies the name of a transcoding template group.
+
+* `name` - (Required, String) Specifies the name of a transcoding template.
```
Comment (niuzhenguo) [inline docs/resources/mpc_transcoding_template_group.md]: duplicated `name` parameter as L58

File: docs/resources/mpc_transcoding_template_group.md [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+  + **1**: HLS
+  + **2**: DASH
+  + **3**: HLS+DASH
+  + **4**: MP4
+  + **5**: MP3
+  + **6**: ADTS
+
+-> If pack_type is set to 5 or 6, do not set video parameters.
+
+* `pvc` - (Optional, Bool) Specifies Whether to enable low bitrate HD. The default value is false.
```
Comment (niuzhenguo) [inline docs/resources/mpc_transcoding_template_group.md]: ditto.

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/mpc' TESTARGS='-run TestAcc'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mpc -v -run TestAcc -timeout 360m -parallel 4
=== RUN   TestAccTranscodingTemplateGroup_basic
=== PAUSE TestAccTranscodingTemplateGroup_basic
=== RUN   TestAccTranscodingTemplate_basic
=== PAUSE TestAccTranscodingTemplate_basic
=== CONT  TestAccTranscodingTemplateGroup_basic
=== CONT  TestAccTranscodingTemplate_basic
--- PASS: TestAccTranscodingTemplate_basic (22.31s)
--- PASS: TestAccTranscodingTemplateGroup_basic (25.59s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mpc       25.716s
```

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/mpc' TESTARGS='-run TestAcc'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mpc -v -run TestAcc -timeout 360m -parallel 4
=== RUN   TestAccTranscodingTemplateGroup_basic
=== PAUSE TestAccTranscodingTemplateGroup_basic
=== RUN   TestAccTranscodingTemplate_basic
=== PAUSE TestAccTranscodingTemplate_basic
=== CONT  TestAccTranscodingTemplateGroup_basic
=== CONT  TestAccTranscodingTemplate_basic
--- PASS: TestAccTranscodingTemplate_basic (22.31s)
--- PASS: TestAccTranscodingTemplateGroup_basic (25.59s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mpc       25.716s
```

File: docs/resources/mpc_transcoding_template_group.md (hunk 11 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template_group.md]: fixed

File: docs/resources/mpc_transcoding_template.md (hunk 10 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template.md]: fixed

File: docs/resources/mpc_transcoding_template_group.md (hunk 12 above)
Comment (Jason-Zhang9309) [reply docs/resources/mpc_transcoding_template_group.md]: fixed

File: huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+	}
+
+	id, err := strconv.ParseInt(d.Id(), 10, 32)
+	if err != nil {
+		diag.FromErr(err)
+	}
+
+	_, err = client.DeleteTemplate(&mpc.DeleteTemplateRequest{TemplateId: id})
+	if err != nil {
+		return fmtp.DiagErrorf("error deleting MPC transcoding template: %s", err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: please use diag.Errorf instead

Comment (Jason-Zhang9309) [reply huaweicloud/services/mpc/resource_huaweicloud_mpc_transcoding_template.go]: sure, fixed

**Comment by Jason-Zhang9309** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/mpc' TESTARGS='-run TestAcc'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/mpc -v -run TestAcc -timeout 360m -parallel 4
=== RUN   TestAccTranscodingTemplateGroup_basic
=== PAUSE TestAccTranscodingTemplateGroup_basic
=== RUN   TestAccTranscodingTemplate_basic
=== PAUSE TestAccTranscodingTemplate_basic
=== CONT  TestAccTranscodingTemplateGroup_basic
=== CONT  TestAccTranscodingTemplate_basic
--- PASS: TestAccTranscodingTemplate_basic (21.98s)
--- PASS: TestAccTranscodingTemplateGroup_basic (26.02s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/mpc       26.125s
```

**Pull request merged**